dynamic_programming/longest_common_string.cpp

```diff
@@ -1,53 +1,111 @@
-        for (j = 0; j < n; j++) {
-            if (a[i][j] > ma) {
-                ma = a[i][j];
-                indi = i;
-                indj = j;
+    const auto limit = static_cast<std::size_t>(-1);
+    for (std::size_t pos_a = size_a - 1; pos_a != limit; --pos_a) {
+        for (std::size_t pos_b = size_b - 1; pos_b != limit; --pos_b) {
+            if (str_a[pos_a] == str_b[pos_b]) {
+                sub_sols[pos_a][pos_b] = 1 + sub_sols[pos_a + 1][pos_b + 1];
+            } else {
+                sub_sols[pos_a][pos_b] = std::max(sub_sols[pos_a + 1][pos_b],
+                                                  sub_sols[pos_a][pos_b + 1]);
             }
         }
     }
 
-    cout << str1 << "\n";
-    cout << str2 << "\n";
+    return sub_sols[0][0];
+}
+
+struct TestCase {
+    const std::string str_a;
+    const std::string str_b;
+    const std::size_t common_string_len;
+
+    TestCase(std::string in_str_a, std::string in_str_b,
+             const std::size_t in_common_string_len)
+        : str_a(std::move(in_str_a)),
+          str_b(std::move(in_str_b)),
+          common_string_len(in_common_string_len) {}
+};
+
+std::vector<TestCase> get_test_cases() {
+    return {TestCase("", "", 0),
+            TestCase("ab", "ab", 2),
+            TestCase("ab", "ba", 1),
+            TestCase("", "xyz", 0),
+            TestCase("abcde", "ace", 3),
+            TestCase("BADANA", "ANADA", 3),
+            TestCase("BADANA", "CANADAS", 3),
+            TestCase("a1a234a5aaaa6", "A1AAAA234AAA56AAAAA", 6),
+            TestCase("123x", "123", 3),
+            TestCase("12x3x", "123", 3),
+            TestCase("1x2x3x", "123", 3),
+            TestCase("x1x2x3x", "123", 3),
+            TestCase("x12x3x", "123", 3)};
+}
+
+void test_longest_common_string_length() {
+    for (const auto& cur_tc : get_test_cases()) {
+        assert(longest_common_string_length(cur_tc.str_a, cur_tc.str_b) ==
+               cur_tc.common_string_len);
+    }
+}
+
+void test_longest_common_string_length_is_symmetric() {
+    for (const auto& cur_tc : get_test_cases()) {
+        assert(longest_common_string_length(cur_tc.str_b, cur_tc.str_a) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief reverses a given string
+ * @param in_str input string
+ * @return the string in which the characters appear in the reversed order as in
+ * in_str
+ */
+std::string reverse_str(const std::string& in_str) {
+    return {in_str.rbegin(), in_str.rend()};
+}
 
-    cout << "longest string size = " << ma /*<<" "<<indi<<" "<<indj*/ << "\n";
-    for (i = indi - 3; i < indi; i++) cout << str1[i];
-    cout << "\n";
+void test_longest_common_string_length_for_reversed_inputs() {
+    for (const auto& cur_tc : get_test_cases()) {
+        assert(longest_common_string_length(reverse_str(cur_tc.str_a),
+                                            reverse_str(cur_tc.str_b)) ==
+               cur_tc.common_string_len);
+    }
+}
+
```

Comment:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
```

Reply:
What is an advantage of having this comment? Is it any easier to read than the code it self, which after applying [your remark](https://github.com/TheAlgorithms/C-Plus-Plus/pull/2462#discussion_r1202245229), looks somehow like this:
```cpp
int main() {
    test_all();
    return 0;
}
```
I would also argue that the comment is not fully true: what does the `main` function return, when some of the tests fail?

Reply:
not test_all()
but tests();

Reply:
this comment is important as it is part of all documentation, our guidelines enforce it so you must follow it

Reply:
> this comment is important as it is part of all documentation, our guidelines enforce it so you must follow it

No it is not. Please have a look at the first _bullet point_ in the [guidelines about the documentation](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#documentation): _Make sure you put useful comments in your code. Do not comment on obvious things._.